text/3446-store.md

```diff
@@ -0,0 +1,1225 @@
+    /// invalid
+    fn dangling(&self, alignment: Alignment) -> Result<Self::Handle, AllocError>;
+}
+
+/// Allocates and deallocates handles to blocks of memory, which can be temporarily resolved to pointers to actually
+/// access said memory.
+pub unsafe trait Store: StoreDangling {
+    /// Return a pointer to the block of memory associated to `handle`.
+    unsafe fn resolve(&self, handle: Self::Handle) -> NonNull<u8>;
+
+    //  The following methods are similar to `Allocator`, reformulated in terms of `Self::Handle`.
+
+    /// Allocates a new handle to a block of memory. On success, invalidates any existing handle.
+    fn allocate(&self, layout: Layout) -> Result<(Self::Handle, usize), AllocError>;
+
+    /// Deallocates a handle.
+    unsafe fn deallocate(&self, handle: Self::Handle, layout: Layout);
+
+    /// Grows the block of memory associated to a handle. On success, the handle is invalidated.
+    unsafe fn grow(
+        &self,
+        handle: Self::Handle,
+        old_layout: Layout,
+        new_layout: Layout,
+    ) -> Result<(Self::Handle, usize), AllocError>;
+
+    /// Shrinks the block of memory associated to a handle. On success, the handle is invalidated.
+    unsafe fn shrink(
+        &self,
+        handle: Self::Handle,
+        old_layout: Layout,
+        new_layout: Layout,
+    ) -> Result<(Self::Handle, usize), AllocError>;
+
+    /// Allocates a new handle to a block of zeroed memory. On success, invalidates any existing handle.
+    fn allocate_zeroed(&self, layout: Layout) -> Result<(Self::Handle, usize), AllocError> {
+        ...
+    }
+
+    /// Grows the block of memory associated to a handle with zeroed memory. On success, the handle is invalidated.
+    unsafe fn grow_zeroed(
+        &self,
+        handle: Self::Handle,
+        old_layout: Layout,
+        new_layout: Layout,
+    ) -> Result<(Self::Handle, usize), AllocError> {
+        ...
+    }
+}
+```
+
+_Note:  full-featured documentation of the trait and methods can be found in the companion repository at_
+        https://github.com/matthieu-m/store/blob/main/src/interface.rs.
+
+The `Store` trait is supplemented by 3 additional marker traits, providing extra guarantees:
+
+```rust
+/// A refinement of `Store` which does not invalidate existing handles on allocation, and does not invalidate
+/// unrelated existing handles on deallocation.
+pub unsafe trait StoreMultiple: Store {}
+
+/// A refinement of `Store` which does not invalidate existing pointers on allocation, resolution, or deallocation, but
+/// may invalidate them on moves.
+pub unsafe trait StoreStable: Store {}
+
+/// A refinement of `Store` which does not invalidate existing pointers, not even on moves. That is, this refinement
+/// guarantees that the blocks of memory are pinned in memory until the instance of the `Store` is dropped, or until
+/// the lifetime bound of `Store` concrete type expires, whichever comes first.
```

Comment:
`Pin` is not going to be relaxed with a lifetime carve out on the `Drop` requirement. Scoped async tasks are unsound because of the ability to borrow from outside of the current task, but would be sound if restricted somehow to borrowing within the same task, specifically because of the drop-before-reuse guarantee. If you remove that guarantee by "pinning" in storage that can be forgotten without leaking, actual extant code becomes unsound.

That this requirement is not strong enough to meet the pinning guarantee without `S: 'static` is a problematic footgun, but that's due to the complexity of the actual strong pinning guarantee and lack of better available name (since this also weakens "stable" to not mean stable across moves) rather than an intent to weaken it.

Mostly, IIRC, due to the use of non-owning storage slipping @matthieu-m's consideration for the initial drafting rounds.

Reply:
> Mostly, IIRC, due to the use of non-owning storage slipping @matthieu-m's consideration for the initial drafting rounds.

I completely forgot about the usecase initially, yes.

The above is actually my attempt at tightening the wording to make `Pin` work. Is it still inaccurate?

Reply:
@CAD97 chewing your response thoroughly:

> Scoped async tasks are unsound because of the ability to borrow from outside of the current task

I'm not really here to talk about scoped async tasks, but to be precise async tasks that are not `'static` may only borrow from a thread that is blocked and will not unblock until the task is either completed or canceled. Borrowing from an unblocked thread is unsound because panics may drop the borrowed values. Unblocking before the task is completed or canceled also provides an opportunity to panic before the future is dropped.

> but would be sound if restricted somehow to borrowing within the same task

Agreed.

> specifically because of the drop-before-reuse guarantee

Borrowing within the same task is sound regardless of whether you have a drop-before-reuse guarantee. The future and all of its constituent parts must be dropped or forgotten together. If they're all dropped, great! If they're all forgotten, then the borrowed data is forgotten at the same time as the borrows that use them.

> If you remove that guarantee by "pinning" in storage that can be forgotten without leaking

To make my position clear, I think I should be able to pin in non-`'static` storage. This is equivalent to pinning in storage that can be "forgotten without leaking" but I'd like to be more precise.

The current pinning guarantees require that allocators return pointers to memory that is not reused unless all of the allocators for it are dropped. Since this depends on a drop guarantee and not a lifetime guarantee, forgetting an allocator must permanently leak the memory it allocated.

These guarantees are stronger than necessary for self-referential types. If a self-referential type is completely forgotten or dropped, then no references are invalidated since the only references were also forgotten or dropped. The strong pinning guarantees are useful for structures like intrusive doubly-linked lists, which _do_ need drop to be called before memory can be reused - forgetting is not sufficient.

For self-referential types, the current guarantees of `StorePinning` are sufficient. For inter-referential types, bounding `S: StorePinning + 'static` is sufficient (as if the storage is forgotten, then its memory is valid for `'static`). I think the way these two use cases are reflected in the additive bounds is elegant and desirable.

> actual extant code becomes unsound.

I am aware that changing the pinning guarantees is very problematic. I just want to know whether those conversations are happening, as I have a vested interest.

Reply:
@CAD97 @matthieu-m I'll note that this issue seems very related to the one raised in a thread on [&move semantics](https://internals.rust-lang.org/t/a-sketch-for-move-semantics/18632). I still believe that the current Pin guarantees are unreasonably strong, preventing its smooth compatibility with &move or Inline Storages, though I don't know how to square the circle of current guarantees.